cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -739,7 +723,15 @@ func (p *BicepProvider) deployModule(ctx context.Context, scope infra.Scope, bic
 	// `DeploymentNotFound` error.
 	// Since other commands of ours use the deployment, let's try to fetch it here and if we fail with `DeploymentNotFound`,
 	// ignore this error, wait a short while and retry.
-	if err := scope.Deploy(ctx, bicepPath, parametersPath); err != nil {
+
+	// deployments API takes an ARM template.
+	// At this point, the bicep file should have been already compiled and succeeded
+	// do panic if the application tries to deploy a bicep file without compiling it first
+	if p.compiledBicep == nil {
+		log.Panic("bicep file was not previously compiled. A call to State() is expected before calling Deploy")
+	}
```

Comment:
The `State()` method is not guaranteed to run before a deployment and is actually run after a deployment.  If we don't have the compiledBicep at this point I would consider just calling the method that you need.

Do we need to introduce this internal component state here and just let it work like it was??

Reply:
I also don't love the introduction of this `compiledBicep` state.  I love the rationale; since I agree there are call paths where we definitely were calling bicep multiple times but I think we should address that outside of these changes.

One thing you might consider is using the `BicepDeploymentDetails` object that is returned by `Plan` to also include the compiled template, and then in `Deploy` pull the compiled template off it.

That might get us most of the way to where we want to be and not require us to add state to the provider itself?

Reply:
updated to use `BicepDeploymentDetails`